### PR TITLE
Updating manure excretions output data passed to OM

### LINKED ISSUE #904: Incorrect number of rows in the aggregate heifer manure outputs
## Expected Behavior

The same number of rows in the aggregate heifer manure outputs compared to other animal categories. 

## Current Behavior

The number of rows for heifers are twice of that of other animal categories. Those integers do not seem to belong there. 

Screenshot 2023-11-03 at 10 48 00 AM

## Possible Solution

Start with investigating how aggregate heifer manure outputs are collected by the output manager.  

## Steps to Reproduce

1. Use the filter 'Pen.calc_total_manure'. 
2.
3.
4.

### PR DESCRIPTION
**Pull request by ltp23**
The PR aims to fix the git issue #904, which is about the inconsistent lengths of csv output columns related to manure excretions in the animal module, specifically for heifers. This happens when in an animal manager's daily update there are multiple pens with overlapping animal groups, the daily update's function call will add the corresponding number of duplicates. 

We happened to observe this with the heifer group, but I think had we had multiple pens of lactating cows, we would have probably seen something similar.

With the current fix, I temporarily store the per-animal-group manure excretion output across all the pen's calculations and then dump the data only once at the end. For the per-pen manure excretion output, this has not changed and still gets evaluated as we iterate over each pen.

The following filter can be used for this new update.
```
.*AnimalManager.daily_updates.daily_aggregate.*
.*Pen.calc_total_manure.*
```

### CODE REVIEW

**Review by tomhuhh — APPROVED** [review]
LGTM! I tested with the filter Loi provided and the outputs are of the same length across different animal categories. Thanks @ltp23 for your hard work!

File: RUFAS/routines/animal/pen.py [hunk 1]
```diff
@@ -422,7 +422,7 @@ def update_animals(self, new_animals: List[Any],
         self.update_animal_combination(animal_combination)
         self.update_classes_in_pen()
 
-    def calc_manure(self, feed, methane_model): # noqa
+    def calc_manure(self, feed, methane_model):  # noqa
         """
         Calculates the total manure excretion of the animals in the pen,
```
Comment (Isgotkowitz) [inline RUFAS/routines/animal/pen.py]: Can this docstring be updated to to NumPy style and the method signature type annotated.

File: RUFAS/routines/animal/pen.py [hunk 2]
```diff
@@ -550,7 +550,7 @@ def daily_p_update(self):
             total_p_animal = max(total_p_animal, 0)
             self.avg_p_animal = total_p_animal / len(self.animals_in_pen)
 
-    def set_up_new_animal(self, animal, p_conc, feed, temp, num_animals_before_additions): # noqa
+    def set_up_new_animal(self, animal, p_conc, feed, temp, num_animals_before_additions):  # noqa
```
Comment (Isgotkowitz) [inline RUFAS/routines/animal/pen.py:555]: Same here as above, can the method signature be type annotated and the docstring style changed to NumPy

**Review by Isgotkowitz — APPROVED** [review]
Code looks good, especially the testing. Two comments, both related to updating docstrings.

While looking through the output, I did have some questions:
- Why do pen specific manure excretion values not distinguish between the animal types that produced them while the aggregate does? Is it assumed that a single pen will only have a single type of cow in them?
- There are negative initial values for some aggregate lactating cow excretion values that seem like they should not be there. This includes the output variables:
  - AnimalManager.daily_updates.daily_aggregate_lactating_cow_urine
  - AnimalManager.daily_updates.daily_aggregate_lactating_cow_manure_mass
  - AnimalManager.daily_updates.daily_aggregate_lactating_cow_total_solids
  - AnimalManager.daily_updates.daily_aggregate_lactating_cow_degradable_volatile_solids
  - AnimalManager.daily_updates.daily_aggregate_lactating_cow_non_degradable_volatile_solids
  - AnimalManager.daily_updates.daily_aggregate_lactating_cow_potassium
Is this a known issue?

**Comment by ltp23** [comment]
> * Why do pen specific manure excretion values not distinguish between the animal types that produced them while the aggregate does? Is it assumed that a single pen will only have a single type of cow in them?

Probably because of the current usage from the SMEs' perspective. Also, one of the complicating factors is the classification of animals as they mature through different stages and change pens. If there is a need for this approach, we can make that happen too.

> * There are negative initial values for some aggregate lactating cow excretion values that seem like they should not be there. This includes the output variables:
>   
>   * AnimalManager.daily_updates.daily_aggregate_lactating_cow_urine
>   * AnimalManager.daily_updates.daily_aggregate_lactating_cow_manure_mass
>   * AnimalManager.daily_updates.daily_aggregate_lactating_cow_total_solids
>   * AnimalManager.daily_updates.daily_aggregate_lactating_cow_degradable_volatile_solids
>   * AnimalManager.daily_updates.daily_aggregate_lactating_cow_non_degradable_volatile_solids
>   * AnimalManager.daily_updates.daily_aggregate_lactating_cow_potassium
>     Is this a known issue?

I think so. We have an open issue #609 that describes this. The negative values come from the unbalance between the calculated values from ration calculations and the formulas we use. However, one temporary, OK fix is to make sure each value is at least non-negative.

File: RUFAS/routines/animal/pen.py (hunk 1 above)
Comment (ltp23) [reply RUFAS/routines/animal/pen.py]: I have updated the docstring for this method, but it needs to be refactored later.

File: RUFAS/routines/animal/pen.py (hunk 2 above)
Comment (ltp23) [reply RUFAS/routines/animal/pen.py:555]: I will hold off on this one because it will probably be removed during the animal manager refactoring.

**Comment by ltp23** [comment]
Thanks Haowen and Ed!!

**Pull request merged**